test/req/steps_test.exs

```diff
@@ -698,6 +698,47 @@ defmodule Req.StepsTest do
     assert Agent.get(:counter, & &1) == 3
   end
 
+  @tag :capture_log
+  test "retry: retry-after", c do
```

Comment:
we could set retry_delay to a really high value and expect it not to be used, otherwise our test would time out. To keep the test super fast we can set retry-after to be 0:

```elixir
  @tag :capture_log
  test "retry: retry-after" do
    adapter = fn request ->
      response =
        case Map.get(request.private, :req_retry_count, 0) do
          0 -> Req.Response.new(status: 429) |> retry_after(0)
          1 -> Req.Response.new(status: 429) |> retry_after(DateTime.utc_now())
          2 -> Req.Response.new(status: 200, body: "ok")
        end

      {request, response}
    end

    assert Req.request!(adapter: adapter, retry_delay: 10000).body == "ok"
  end

  defp retry_after(r, value), do: Req.Response.put_header(r, "retry-after", retry_after(value))
  defp retry_after(integer) when is_integer(integer), do: to_string(integer)
  defp retry_after(%DateTime{} = dt), do: Req.Steps.format_http_datetime(dt)
```

we'd just need to make our slight change in production code, something like:

```diff
  delay
  |> ...
  |> DateTime.diff(DateTime.utc_now(), :millisecond)
+ |> max(0)
```

wdyt?

Reply:
I like it! There are a few things I wonder about:

1. Depending on the internal details of retry counting
2. We don't really know anything about the delay that was/was not applied except by virtue of watching the test suite run and seeing that it wasn't really slow.

I'll work on my suggestion for addressing those two items.

Reply:
Both are valid concerns. Regarding 1, yeah, if internals change, the test will have to change too but I'm fine with that and fwiw the test will spectacularly fail so I'm not worried about not catching this. Regarding 2, besides noticing it is really slow, ExUnit will crash the test when it takes too long. The default timeout is 60s so my proposed delay was way too slow. We'd either do way bigger delay or `@tag timeout: 1000`.

Reply:
Oh, btw, the beauty (if I may say so) of using custom adapter is we could do this:

```elixir
    adapter = fn request ->
      attempt = Req.Request.get_private(request, :attempt, 0)

      response =
        case attempt do
          0 -> Req.Response.new(status: 429) |> retry_after(0)
          1 -> Req.Response.new(status: 429) |> retry_after(DateTime.utc_now())
          2 -> Req.Response.new(status: 200, body: "ok")
        end

      request = Req.Request.put_private(request, :attempt, attempt + 1)
      {request, response}
    end
```

and it would totally work too. Now, we'd still be coupled to how retry/1 works, that it re-uses request struct, but that's something way more stable. In fact if this sounds good to you why we don't ship with this? I guess We can skip send/assert_receive ping too, I don't think it adds much after all.

Reply:
I just added `update_private` so it can be even simpler:

```elixir
adapter = fn request ->
  request = Req.Request.update_private(request, :attempt, 0, & &1 + 1)
  attempt = request.private.attempt
  # ...
```